File: processing/src/server/ColorMatcher.java
package server;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.LinkedList;
import java.util.List;
import java.util.Map.Entry;

public class ColorMatcher {

	private static final int QUEUE_LENGTH = 10;

	private static final long MAX_TIME = 1000L;

	private static final boolean SCANNING_MODE = false;

	private long timeLastMatch = 0;

	boolean started = false;

	/** when adding colors to queue, interpolate color if distance over this value */
	private float maxColorDistance = 0.1f;
	public float matchDistance = 0.06f;

	/** color buffer that hold the last detected colors */
	private LinkedList<YUV> colors = new LinkedList<YUV>();

	ColorMatchListener listener;
	
	// fields for scanning the neutral color
	private long scanStart = 0;
	private YUV neutral = null;
	private YUV averagedColor = null;
	private List<YUV> scannedColors = new ArrayList<YUV>();
	private float scanDistance = 0.1f;

	public interface ColorMatchListener {

		public void onColorMatch(List<YUV> colors);

	}

	public static HashMap<YUV, String> colorMap;

	static {
		colorMap = new HashMap<YUV, String>();
		colorMap.put(new YUV(180, 58, 50), "meat");
		colorMap.put(new YUV(75, 113, 62), "vegetables");
		colorMap.put(new YUV(125, 87, 35), "carbs");
		colorMap.put(new YUV(63, 127, 52), "light-green");
	}


	private YUV previouslyMatched = null;

	/** contains all matched colors including neutral values */
	private List<YUV> matched = new ArrayList<YUV>();
	
	/** includes all matched colors without neutral values */
	private List<YUV> matchedStripped = new ArrayList<YUV>();

	public ColorMatcher(ColorMatchListener listener){
		this.listener = listener;
	}

	/**
	 * Match a color with a color from the color table {@link #colorMap}.
	 * If a match is detected (for the first time), the interface function {@link #ColorMatchListener)}
	 * is called
	 * 
	 * @param color
	 */
	public void match(YUV color){
		if(SCANNING_MODE){
			scanColor(color);
		} else {
			// detect neutral color first
			if(neutral == null){
				findNeutralColor(color);
				return;
			}
			
			// if neutral is found then proceed with matching colors
			addColorToQueue(color);

			YUV bestMatch = null;

			for (Entry<YUV, String> entry : colorMap.entrySet()) {
				if(color.distanceTo(entry.getKey()) < matchDistance){
					if(bestMatch == null){
						bestMatch = entry.getKey();
					} else {
						if(color.distanceTo(entry.getKey()) < 
								color.distanceTo(bestMatch)){
							bestMatch = entry.getKey();
						}
					}		
				}
			}

			if(bestMatch != null && bestMatch != previouslyMatched){
				if(timeLastMatch == 0 || System.currentTimeMillis() - timeLastMatch < MAX_TIME){
					previouslyMatched = bestMatch;
					timeLastMatch = System.currentTimeMillis();
					matched.add(bestMatch);
				}
			}
		}
	}

	private void scanColor(YUV color) {
		if(scanStart == 0){
			scanStart = System.currentTimeMillis();
		}

		// find neutral color
		if(System.currentTimeMillis() - scanStart < 3000){
			if(neutral == null){
				neutral = color.copy();
			} else {
				neutral = color.average(neutral);
			}
		} else {
			if(color.distanceTo(neutral) > scanDistance){
				scannedColors.add(color.copy());

				float y = 0;
				float u = 0;
				float v = 0;
				int n = 0;;
				for (YUV c : scannedColors) {
					y += c.y;
					u += c.u;
					v += c.v;
					n++;
				}

				averagedColor = new YUV(y / (float)n, u / (float)n, v / (float)n, true);

				System.out.println("Scanned: " +color.getRGB().toString());
				System.out.println("Averaged: " +averagedColor.getRGB().toString());
			}
		}
	}

	private void findNeutralColor(YUV color){
		if(scanStart == 0){
			scanStart = System.currentTimeMillis();
		}

		// find neutral color
		if(System.currentTimeMillis() - scanStart < 3000){
			if(neutral == null){
				neutral = color.copy();
			} else {
				neutral = color.average(neutral);
			}
		} else {
			colorMap.put(neutral, "neutral");
		}
	}

	public void checkMatchedColors(){
		if(matched.size() > 0 && System.currentTimeMillis() - timeLastMatch > MAX_TIME){
			matchedStripped.clear();
			
			// remove all neutral colors from matched list
			for (YUV matchedColor : matched) {
				if(!matchedColor.equals(neutral)){
					matchedStripped.add(matchedColor);
				}
			}
			
			listener.onColorMatch(matchedStripped);
			matched.clear();
			timeLastMatch = 0;
			previouslyMatched = null;
		}
	}

	public void addColorToQueue(YUV color){
		YUV averagedColor = null;

		// interpolate color if distance is too
		if(colors.size() > 0){
			if(color.distanceTo(colors.peek()) > maxColorDistance){
				averagedColor = color.average(colors.peek());
				System.out.println("averaged color");
			}
		}

		if(colors.size() < QUEUE_LENGTH){
			if(averagedColor != null){
				colors.push(averagedColor);
			} else {
				colors.push(color);
			}
		} else {
			colors.remove(colors.getLast());
			if(averagedColor != null){
				colors.push(averagedColor);
			} else {
				colors.push(color);
			}
		}
	}
}

File: python/recipe-printer-server/colormatcher.py
#!/usr/bin/env python2

import time
import math
from YUV import *
from RGB import *

class ColorMatcher:
    MAX_TIME = 1 # max waiting time for scanner
    
    SCANNING_MODE = False # set true to scan marbles

    timeLastMatch = 0 # last time a match was found

    # maxColorDistance = 0.1; # not needed; 
    matchDistance = 0.04; # color distance to color in the colormap

    # color buffer that hold the last detected colors */
    colors = [];
    
    # fields for scanning
    scanStart = 0;
    neutral = None; # neutral YUV color
    averagedColor = None; # neutral YUV color
    scannedColors = [];
    scanDistance = 0.1;
    
    # colormap holding predefined colors to match
    colorMap = dict({
        YUV.fromIntRgb(138, 64, 52) : "meat",
        YUV.fromIntRgb(89, 94, 61) : "vegetables",
        YUV.fromIntRgb(115, 85, 39) : "carbs",
        YUV.fromIntRgb(69, 85, 88) : "person"
    })

    # previously matched YUV color
    previouslyMatched = None;

    # contains all matched colors including neutral values: YUV
    matched = []
    
    # includes all matched colors without neutral values: YUV
    matchedStripped = [];
    
    def match(self, color):
        if self.SCANNING_MODE:
            self.scanColor(color)
        else:
            # detect neutral color first
            if not "neutral" in self.colorMap.values():
                # find neutral color; this takes 3000 ms
                self.findNeutralColor(color);
                return;

            # if neutral is found then proceed with matching colors
            #addColorToQueue(color);

            bestMatch = None; # best color match: YUV

            # loop over color map to find matches
            for targetColor in self.colorMap:
                # see if the distance from the scanned color to the colorMap is small enough
		if color.distanceTo(targetColor) < self.matchDistance:
                    if bestMatch == None:
                        bestMatch = targetColor
                    else: # if we have multiple matches, only take the best one (smalles distance)
                        if color.distanceTo(targetColor) < color.distanceTo(bestMatch):
                            bestMatch = targetColor
    
                # only match color if it was not recognized before and max scanning time was not reached
                # since we also detect neutral color a list could look like this:
                # red - neutral - red - neutral - green
                if bestMatch != None and bestMatch != self.previouslyMatched:
                    # only add color to the matched colors list if the last matching time was not
                    # more than MAX_TIME ago
                    if self.timeLastMatch == 0 or int(time.time()) - self.timeLastMatch < self.MAX_TIME:
                        self.previouslyMatched = bestMatch
                        self.timeLastMatch = int(time.time()) # save time of this match
                        self.matched.append(bestMatch)
                        print("matched")
			for color in self.matched:
			    print("matched color:" + str(self.colorMap.get(bestMatch)))

    # scan a single YUV color
    # the first 1000 ms of the scanning process the neutral color is found
    # Insert the marble multiple times to form an average value
    def scanColor(self, color):
	print("in scanColor, y:" + str(color.y) + " u:"+ str(color.u) + " v:"+ str(color.v))
	print("scanstart:" + str(self.scanStart) + " timediff:" + str(time.time() - self.scanStart))
        if self.scanStart == 0:
            self.scanStart = int(time.time())
            
        elif int(time.time()) - self.scanStart < 1:
            if self.neutral == None:
                self.neutral = color.copy()
            else:
                self.neutral = color.average(self.neutral)
        else:
            if color.distanceTo(self.neutral) > self.scanDistance:
                self.scannedColors.append(color.copy())
                
                y = 0.0
                u = 0.0
                v = 0.0
                n = 0.0
            
                for c in self.scannedColors:
                    y += c.y
                    u += c.u
                    v += c.v
                    n += 1.0
                
                if self.averagedColor == None:
                    self.averagedColor = YUV(0,0,0)
                
                self.averagedColor = YUV(y / n, u / n, v / n)  
            
                print("Scanned: " +color.getRGB().toString())
                print("Averaged: " + self.averagedColor.getRGB().toString())

    # find the neutral color: YUV
    # takes 3000 ms
    def findNeutralColor(self, color):
	if self.scanStart == 0:
            self.scanStart = int(time.time())
            
        if int(time.time()) - self.scanStart < 1: # average neutral color
            if self.neutral == None:
                self.neutral = color.copy()
            else:
                self.neutral = color.average(self.neutral)
        else:
            self.colorMap[self.neutral] = "neutral" # add neutral color to colormap

    # call this function periodically to check if a scanning process is over
    # returns null if no scanning is done and no colors are detected
    # returns all matched colors with the neutral color filtered out
    def checkMatchedColors(self):
        if len(self.matched) > 0 and int(time.time()) - self.timeLastMatch > self.MAX_TIME:
            self.matchedStripped[:] = []
            
            #remove all neutral colors from matched list
            for c in self.matched:
                if c.equals(self.neutral):
                    self.matchedStripped.append(c)
            
            result = self.matched
            
            # reset values
            self.matched[:] = []
            self.timeLastMatch = 0
            self.previouslyMatched = None
            return result
        else:
            return None
